### ISSUE TITLE
Populate does not consistently supply default value for array of subdocs.

### ISSUE THREAD

**Issue by wilimitis** [Automattic/mongoose#6245]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
`populate` does supply the default value, `[]`, for array fields on subdocuments when the first subdocument in an array of subdocuments does not have the field in question set (sorry if this is not the minimal representation of the issue, I attempted to illustrate through the repro steps). For instance, the field may not be set if the schema was updated to contain the field in question.

**If the current behavior is a bug, please provide the steps to reproduce.**
The field under test here is `DepartmentSchema.employees`.

1. Create a `Company` with a `department` that does not have `employees` set (here, neither "Engineering" nor "Security" has `employees`, as the field is not yet on `DepartmentSchema`).
```
'use strict';
const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');

const DepartmentSchema = new mongoose.Schema({name: String});
const CompanySchema = new mongoose.Schema({name: String, departments: [DepartmentSchema]});

const Company = mongoose.model('Company', CompanySchema);
const company = new Company({
  name: 'Uber',
  departments: [{name: 'Security'}, {name: 'Engineering'}],
});
company.save();
```
2. Add the `employees` field to `DepartmentSchema`, and add an `employee` to the **first** `department` of the `company`. Notice the default value is supplied for `employees`.
```
'use strict';
const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');

const EmployeeSchema = new mongoose.Schema({name: String});
const DepartmentSchema = new mongoose.Schema({name: String, employees: [EmployeeSchema]});
const CompanySchema = new mongoose.Schema({name: String, departments: [DepartmentSchema]});

const Company = mongoose.model('Company', CompanySchema);
Company.findOne({name: 'Uber'})
  .then(async (uber) => {
    let engineering = uber.departments.find((d) => d.name === 'Security');
    engineering.employees.addToSet({name: 'Kurt'});
    await uber.save();
    let {departments} = await Company.findOne({name: 'Uber'})
      .populate('departments.employees');
    console.log(departments[1].name); // Engineering
    console.log(departments[1].employees); // [] <--- default value
  });
```
3. Remove the record created in step 1, run step 1 again.
4. Add the `employees` field to `DepartmentSchema`, and add an `employee` to the **second** `department` of the `company`. Notice the default value is not supplied.
```
'use strict';
const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');

const EmployeeSchema = new mongoose.Schema({name: String});
const DepartmentSchema = new mongoose.Schema({name: String, employees: [EmployeeSchema]});
const CompanySchema = new mongoose.Schema({name: String, departments: [DepartmentSchema]});

const Company = mongoose.model('Company', CompanySchema);
Company.findOne({name: 'Uber'})
  .then(async (uber) => {
    let engineering = uber.departments.find((d) => d.name === 'Engineering');
    engineering.employees.addToSet({name: 'Kurt'});
    await uber.save();
    let {departments} = await Company.findOne({name: 'Uber'})
      .populate('departments.employees');
    console.log(departments[0].name); // Security
    console.log(departments[0].employees); // null <--- ???
  });
```

**What is the expected behavior?**
Consistency either way, preferably in favor of returning the default value (ex. step 4 should print `[]`).

**Please mention your node.js, mongoose and MongoDB version.**
node.js: v8.9.3
mongoose: 5.0.10
mongodb: 3.6

**Comment by lineus** [comment]
@wilimitis thanks for making this so easy to reproduce. 
here's the code I used for step 3
```
'use strict'

const mongoose = require('mongoose')

async function run () {
  let collection = 'companies'
  let m = await mongoose.connect('mongodb://localhost/test')
  m.connections[0].dropCollection(collection, (err) => {
    mongoose.connection.close()
    if (err) {
      return console.error(err)
    }
    return console.log(`collection ${collection} dropped.`)
  })
}

run()
```
after changing the console log calls in your original scripts,
here's my output:
```
6245: node ./1.js && node ./2.js && node ./3.js && node ./1.js && node ./4.js 
company added.
  Kurt added to Security.employees
    Engineering Engineering
    Engineering.employees: []
collection companies dropped.
company added.
  Kurt added to Engineering.employees
    Security: Security
    Security.employees: null

6245: 
```

**Comment by vkarpov15** [comment]
Thanks for reporting, will investigate ASAP

**Comment by wilimitis** [comment]
Yes, the call to `populate`, as you've pointed out, is useless and was causing a problem around an unintentional use case - thank you for looking into this!

**Comment by vkarpov15** [comment]
The bug does also affect the case where you have an Employee model that you're populating against:

```javascript
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh-6245`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
mongoose.Promise = global.Promise;

run().catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  await step1();
  delete mongoose.models['Company'];
  await step2();

  console.log('Done');
}

async function step2() { 
  const EmployeeSchema = new mongoose.Schema({name: String}); 
  const DepartmentSchema = new mongoose.Schema({name: String, employees: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Employee' }]});
  const CompanySchema = new mongoose.Schema({name: String, departments: [DepartmentSchema]});
  
  const Employee = mongoose.model('Employee', EmployeeSchema);
  const Company = mongoose.model('Company', CompanySchema);
  const uber = await Company.findOne({name: 'Uber'});
  const kurt = await Employee.create({ name: 'Kurt' });


  let engineering = uber.departments.find((d) => d.name === 'Engineering');
  engineering.employees.addToSet(kurt);
  await uber.save();


  let {departments} = await Company.findOne({name: 'Uber'}).populate('departments.employees');
  console.log(departments[0].name); // Security
  console.log(departments[0].employees); // null <--- ???
}

async function step1() {
  const DepartmentSchema = new mongoose.Schema({name: String});
  const CompanySchema = new mongoose.Schema({name: String, departments: [DepartmentSchema]});

  const Company = mongoose.model('Company', CompanySchema);
  const company = new Company({
    name: 'Uber',
    departments: [{name: 'Security'}, {name: 'Engineering'}],
  });
  await company.save();
}
```

**Comment by vkarpov15** [comment]
Fix will be in 5.0.12
